app/services/appointment_scheduler.rb

```diff
@@ -1,48 +1,115 @@
+          vaccine_name: current_appointment.vaccine_name
+        )
 
-      [:success, appointment]
+        # Free up current appointment if it wasn't checkout out (completed)
+        unless current_appointment.checked_out?
+          current_appointment.update!(patient: nil, check_in: nil, check_out: nil, second_dose: nil, vaccine_name: nil)
+        end
+      end
+
+      [SUCCESS, new_appointment]
     end
-  rescue ActiveRecord::SerializationFailure
-    [:all_slots_taken]
-  rescue => e
-    Sentry.capture_exception(e)
+  end
+
+  # Cancels schedule for an appointment for a given user, in a SQL efficient way
+  def cancel_schedule(patient:, id:)
+    patient.appointments
+           .open
+           .where(id: id)
+           .update_all(patient_id: nil, updated_at: Time.zone.now) # rubocop:disable Rails/SkipsModelValidations
+  end
+
+  def open_times_per_ubs(from:, to:)
+    from = [from, earliest_allowed].compact.max - ROUNDING
+
+    Appointment.active_ubs
+               .active
+               .not_scheduled
+               .open
+               .where(start: rounded_from(from)..to)
+               .order(:start) # in chronological order
+               .select(:ubs_id, :start, :end) # only return what we care with
+               .distinct # remove duplicates (same as .uniq in pure Ruby)
+               .group_by(&:ubs) # transforms it into a Hash grouped by Ubs
+  end
+
+  # Returns how many days ahead there are available appointments
+  def days_ahead_with_open_slot
+    next_available_appointment = Appointment.active_ubs
+                                            .open
+                                            .not_scheduled
+                                            .where(start: earliest_allowed..latest_allowed)
+                                            .order(:start)
+                                            .pick(:start)
+
+    raise NoFreeSlotsAhead unless next_available_appointment
+
+    ((next_available_appointment - Time.zone.now.end_of_day) / 1.day).ceil
+  end
+
+  protected
+
+  # Checks if we can schedule +from+ that time and then rounds to avoid issues with time conversion
+  def rounded_from(from)
+    [earliest_allowed, from].compact.max - ROUNDING
+  end
+
+  def update_appointment(patient_id:, start:, ubs_id:)
+    # Single SQL query to update the first available record it can find
+    # it will either return 0 if no rows could be found, or 1 if it was able to schedule an appointment
+    appointment = Appointment
+                  .active_ubs
+                  .open
+                  .not_scheduled
+                  .order(:start)
+                  .where(start: start)
+                  .limit(1)
+
+    appointment = appointment.where(ubs_id: ubs_id) if ubs_id.present?
 
-    [:internal_error, e.message]
+    appointment
+      .update_all(patient_id: patient_id, updated_at: Time.zone.now) # rubocop:disable Rails/SkipsModelValidations
+      .positive? # 0 = false, 1 = true
```

Comment:
What happens when two or more people request this at the same time? What if there are 3 people and 2 free slots?

Reply:
This `update_all` with `limit(1)` will ensure that only 1 record is updated, and will return the number of rows that are updated. This is why we have a `positive?` check -- it will return `1` (`true`) in case it was able to schedule. If 3 people are trying to get 2 slots, only 2 will schedule and the last one will get it as `false`.